packages/circuit-ui/hooks/useCollapsible/useCollapsible.ts

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  useState,
+  useRef,
+  useCallback,
+  RefObject,
+  KeyboardEvent,
+  MouseEvent,
+} from 'react';
+
+import { uniqueId } from '../../util/id';
+import { useAnimation } from '../useAnimation';
+
+const DEFAULT_HEIGHT = 'auto';
+
+export type CollapsibleOptions = {
+  initialOpen?: boolean;
+  duration?: number;
+  id?: string;
+};
+
+type ButtonProps = {
+  'onClick': (event: MouseEvent | KeyboardEvent) => void;
+  'tabIndex': number;
+  'aria-controls': string;
+  'aria-expanded': 'true' | 'false';
+};
+
+type ContentProps<T> = {
+  'ref': RefObject<T>;
+  'id': string;
+  'style': {
+    overflow: 'hidden';
+    willChange: 'height';
+    opacity: 1 | 0;
+    height: string | 0;
+  };
+  'aria-hidden': null | 'true';
+};
+
+type Collapsible<T> = {
+  isOpen: boolean;
+  toggleOpen: () => void;
+  getButtonProps: (props?: {
+    onClick?: (event: MouseEvent | KeyboardEvent) => void;
+  }) => ButtonProps;
+  getContentProps: (props?: {
+    style?: Record<string, string>;
+  }) => ContentProps<T>;
+};
+
+/**
+ * A hook to build accessible and smoothly animated collapsible sections.
+ * Based on https://inclusive-components.design/collapsible-sections/
+ */
+export function useCollapsible<T extends HTMLElement = HTMLElement>({
+  initialOpen = false,
+  duration = 300,
+  id,
+}: CollapsibleOptions = {}): Collapsible<T> {
+  const detailElement = useRef<T>(null);
+  const [isOpen, setOpen] = useState(initialOpen);
+  const [height, setHeight] = useState(getHeight(detailElement));
+  const [, setAnimating] = useAnimation();
+
+  const detailId = id || uniqueId('collapsible_');
+
+  const toggleOpen = useCallback(() => {
+    setAnimating({
+      duration,
+      onStart: () => {
+        setHeight(getHeight(detailElement));
+        if (!isOpen) {
+          setOpen(true);
+        }
+      },
+      onEnd: () => {
+        if (isOpen) {
+          setOpen(false);
+        }
+        setHeight(DEFAULT_HEIGHT);
+      },
+    });
+  }, [isOpen, detailElement, setAnimating, duration]);
+
+  return {
+    isOpen,
+    toggleOpen,
+    getButtonProps: (props = {}) => ({
+      'onClick': (event: MouseEvent | KeyboardEvent) => {
+        if (props.onClick) {
+          props.onClick(event);
+        }
+        toggleOpen();
+      },
+      'tabIndex': 0,
+      'aria-controls': detailId,
+      'aria-expanded': isOpen ? 'true' : 'false',
+    }),
+    getContentProps: (props = {}) => ({
+      'ref': detailElement,
+      'id': detailId,
+      'style': {
+        overflow: 'hidden',
+        willChange: 'height',
+        opacity: isOpen ? 1 : 0,
+        height: isOpen ? height : 0,
+        visibility: isOpen ? 'visible' : 'hidden',
+        transition: `all ${duration}ms ease-in-out`,
```

Comment:
Should the duration really be customizable, or should it come from the theme? I don't want us to end up with vastly different animation speeds

Reply:
I think so. This is a low-level building block that should be more configurable to enable diverse use cases. It would be simple to standardize this in the future if we change our minds.